distributed/deploy/spec.py

```diff
@@ -328,6 +328,24 @@ def new_worker_spec(self):
 
     scale_up = scale  # backwards compatibility
 
+    def logs(self):
+        async def get_logs(task):
+            log = ""
+            async for line in task.logs():
+                log += "{}\n".format(line)
+            return Log(log)
+
+        scheduler_logs = {
+            "scheduler - {}".format(self.scheduler.address): self.sync(
+                get_logs, self.scheduler
+            )
+        }
+        worker_logs = {
+            "worker {} - {}".format(key, worker.address): self.sync(get_logs, worker)
+            for key, worker in self.workers.items()
+        }
+        return Logs({**scheduler_logs, **worker_logs})
```

Comment:
Typically we try to support both synchronous and asynchronous modes by having two functions:

```python
async def _logs(self):
    ...

def logs(self):
    return self.sync(self._logs)
```

The `self.sync` method takes care to return the async future if we're in an asynchronous context or to run things in the event loop itself, block, and return the result if we're in a synchronous context.

Reply:
We might also provide a default implementation using `self.scheduler_comm.logs` and `self.scheduler_comm.worker_logs`.  This queries the scheduler and workers for logs using Dask communication, but is a bit worse, because it only picks up messages logged to the `distributed` logging handler.

Reply:
I hadn't considered getting logs over the Dask communication. 

I would feel better if the process classes were responsible for providing their own logs on demand by the cluster. Going through the `cluster.scheduler_comm` sidesteps that. Ideally the process classes would speak directly to the resources they have created, but I guess we can't assume that the workers are routable from our Python session.

Reply:
I agree that it would be better to get logs from the resource manager, this will have the advantage of getting things like stdout as well.

I'm just proposing using the Dask methods as a default implementation, in case the whoever implements this class doesn't bother to implement a logs method.

Reply:
Ah gotcha. So instead of the `raise NotImplementedError`.

Reply:
Yes.   That.

Reply:
Ok I'll look at that over in #2874.